### ISSUE TITLE
[Bug?]: Hydration Mismatch during Development. Nesting element within an object will break hydration. 

### ISSUE THREAD

**Issue by edygar** [solidjs/solid-start#1475]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Route throws consistently when accessing it during development, any time the route is accessed.  Production build seems to work fine. Anticipating [this question](https://github.com/solidjs/solid-start/issues/1473#issuecomment-2105257033), it doesn't influences at all, restarting the server doesn't affect it..

![image](https://github.com/solidjs/solid-start/assets/566280/50af420c-ef6d-4777-a1e8-7ac77e83acd7)

### Expected behavior 🤔

It shouldn't throw.

### Steps to reproduce 🕹

Steps:

1.  Create a route with the following code:
```tsx
import type { JSX } from "solid-js";

const ReproduceSteps = (props: {
  test: { propertyBesidesElement: JSX.Element; element: JSX.Element };
}) => {
  return (
    <button onClick={() => console.log(props.test)}>
      {props.test.propertyBesidesElement}
    </button>
  );
};
export default function BugReport() {
  return (
    <ReproduceSteps
      test={{ propertyBesidesElement: "test", element: <div>test</div> }}
    />
  );
}
```

2.  Run the project in `dev` mode
3. Access the route

### Context 🔦

I've reduced it to the minimal reproducible code. Let me know if anything else is required. Couldn't find this same bug on other issues so I decided to report.

### Your environment 🌎

```shell
System:
OS: MacOS Sonoma 14.4.1
CPU: Apple M1 Pro (16 GB of Memory)
Binaries:
Node: v20.10.0 - $HOME/.volta/bin/node
pnpm: 8.14.0 - $HOME/.volta/bin/pnpm
npmPackages:
"@solidjs/meta": "^0.29.3",      
"@solidjs/router": "^0.13.3",    
"@solidjs/start": "^1.0.0-rc.1", 
"solid-js": "^1.8.17",           
"vinxi": "^0.3.11"
```

**Comment by peerreynders** [comment]
Context: [Original Discord discussion](https://discord.com/channels/722131463138705510/1238643925988937820).

TLDR: Feels that requiring a render thunk
```ts
export default function BugReport() {
  return (
    <ReproduceSteps
      test={{ propertyBesidesElement: "test", element: () => <div>test</div> }}
    />
  );
}
```
for SSR is poor DX.

**Comment by peerreynders** [comment]
Context: [Original Discord discussion](https://discord.com/channels/722131463138705510/1238643925988937820).

TLDR: Feels that requiring a render thunk
```ts
import type { JSX } from 'solid-js';

const ReproduceSteps = (props: {
  test: {
    propertyBesidesElement: () => JSX.Element;
    element: () => JSX.Element;
  };
}) => {
  return (
    <button onClick={() => console.log(props.test)}>
      {props.test.propertyBesidesElement()}
    </button>
  );
};

export default function App() {
  return (
    <ReproduceSteps
      test={{
        propertyBesidesElement: () => 'test',
        element: () => <div>test</div>,
      }}
    />
  );
}
```
to support SSR is poor DX.

**Comment by ryansolid** [comment]
The reason is that technically every time you call props.test you are remaking the DOM elements. Some are being inserted and the others aren't, and aren't being found. In the past we were more lenient here creating new DOM elements in these cases and just warning. But this isn't a good pattern anyway as you are basically doing wasted work by rendering twice. If you access props multiple times we recommend using `children`.

**Comment by edygar** [comment]
Got it:

<img width="423" alt="Screenshot 2024-05-14 at 16 24 14" src="https://github.com/solidjs/solid-start/assets/566280/ad417976-c294-4e62-b24a-8c70f26b3183">

vs: 

<img width="423" alt="Screenshot 2024-05-14 at 16 26 29" src="https://github.com/solidjs/solid-start/assets/566280/269be9af-6137-4bac-8f2e-223095f92d0e">

Conclusion: `_tmpl$2( )` fit into the heuristic, compiler turned that property into a lazily computed one, so the object is rebuilt every time the property is accessed, hence a new element is created, hydration mismatch. So, the solution is to simply wrap the `props.test` reading, through a memo (createMemo).
 
But I'm accessing only once in the Entire `ReproduceSteps` server's runtime. Why did it throw anyways?

**Comment by edygar** [comment]
Also, I see this as very risk approach, I don't to be recreating "consumer" elements all the time. Besides @peerreynders's [recommendation](https://github.com/solidjs/solid-start/issues/1475#issuecomment-2110345797), and ["3 Patterns to Write Declarative, More Readable SolidJS Components"](https://raqueebuddinaziz.com/blog/3-patterns-to-write-better-and-more-readable-solidjs-components/)

I intend to contribute onn the documentation, xD to put a very big warning there against this pattern, which I'm very used from React-based codebases.
